### ISSUE TITLE
Ubuntu 17.10 / Wayland Error

### ISSUE THREAD

**Issue by gaites** [tomaka/glutin#949]
I have no idea what I am doing, so if there's an obvious solution, I'm sorry. 

> thread 'main' panicked at 'Library libwayland-egl.so could not be loaded.', /checkout/src/libcore/option.rs:839:4

> echo $WAYLAND_DISPLAY gives wayland-0

Ubuntu 17.10, Gnome 26. setting WAYLAND_DISPLAY=wayland-1 fixes the problem. Both unset and wayland-0 cause the error above.  

I fully expect that there's an obvious solution that I'm missing, any help appreciated. Thanks!

**Comment by vberger** [comment]
Do you have the package `libwayland-egl1-mesa` installed? (assuming you are using the mesa graphic drivers)

**Comment by gaites** [comment]
Here's the output of `sudo apt install libwayland-egl1-mesa`:

> libwayland-egl1-mesa is already the newest version (17.2.2-0ubuntu1).
libwayland-egl1-mesa set to manually installed.
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded. `

and the output of `lshw -c video`:

> description: VGA compatible controller
       product: 2nd Generation Core Processor Family Integrated Graphics Controller
       vendor: Intel Corporation
       physical id: 2
       bus info: pci@0000:00:02.0
       version: 09
       width: 64 bits
       clock: 33MHz
       capabilities: vga_controller bus_master cap_list rom
       configuration: driver=i915 latency=0
       resources: irq:35 memory:c0000000-c03fffff memory:b0000000-bfffffff ioport:6000(size=64) memory:c0000-dffff

I'm not sure if this would make any difference. I do have a hybrid graphics with an older AMD Radeon mobile card, but I'm not running with DRI_PRIME=1.

**Comment by vberger** [comment]
Oh, I suspect I see the issue, which of the following files do you have on your system?

- `/usr/lib/libwayland-egl.so`
- `/usr/lib/libwayland-egl.so.1`
- `/usr/lib/libwayland-egl.so.1.0.0`

**Comment by gaites** [comment]
We've got a winner in another directory:

> /usr/lib/x86_64-linux-gnu/libwayland-egl.so.1
/usr/lib/x86_64-linux-gnu/libwayland-egl.so.1.0.0

Is this a call for a symlink to /lib/usr?

**Comment by vberger** [comment]
No, this is a call to making loading of libwayland-egl in wayland-sys more resiliant. I'll fix this right now, it's a very simple fix.

**Comment by vberger** [comment]
Though simlinking libwayland-egl.so.1 to libwayland-egl.so in the same directory would be a working hotfix, indeed.

**Comment by gaites** [comment]
Glad that's an easy fix! Following your recommendation gets past the original error, resulting in:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: NotSupported', /checkout/src/libcore/result.rs:906:4
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:69
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:58
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:397
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:538
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:522
   7: rust_begin_unwind
             at /checkout/src/libstd/panicking.rs:498
   8: core::panicking::panic_fmt
             at /checkout/src/libcore/panicking.rs:71
   9: core::result::unwrap_failed
             at /checkout/src/libcore/macros.rs:23
  10: <core::result::Result<T, E>>::unwrap
             at /checkout/src/libcore/result.rs:772
  11: clueless::main
             at src/main.rs:18
  12: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:99
  13: std::rt::lang_start
             at /checkout/src/libstd/panicking.rs:459
             at /checkout/src/libstd/panic.rs:361
             at /checkout/src/libstd/rt.rs:59
  14: main
  15: __libc_start_main
  16: _start
```

**Comment by vberger** [comment]
For the record, the fix for not finding the library is here : https://github.com/Smithay/wayland-rs/pull/156

Now, this second error seems like line 18 in your main source file your calling unwrap() on a result that failed, can you paste the associated code here?

**Comment by gaites** [comment]
Here's everything. As you can see this is directly from the README.md with just a couple of small modifications.

```
extern crate glutin;
extern crate gl;
extern crate libc;

use glutin::{GlContext};
use glutin::os::unix::EventsLoopExt;

fn main() {
    let mut events_loop = glutin::EventsLoop::new();
    let window = glutin::WindowBuilder::new()
        .with_title("I'm clueless urghh")
        .with_dimensions(1024, 768);
    let context = glutin::ContextBuilder::new()
        .with_vsync(true);
    let gl_window = glutin::GlWindow::new(window, context, &events_loop).unwrap();

    unsafe {
        gl_window.make_current().unwrap();
    }

    unsafe {
        gl::load_with(|symbol| gl_window.get_proc_address(symbol) as *const _);
        gl::ClearColor(1.0, 1.0, 0.0, 1.0);
    }
    println!("{}", events_loop.is_x11());
    let mut running = true;
    while running {
        events_loop.poll_events(|event| {
            match event {
                glutin::Event::WindowEvent{ event, .. } => match event {
                    glutin::WindowEvent::Closed => running = false,
                    glutin::WindowEvent::Resized(w, h) => gl_window.resize(w, h),
                    glutin::WindowEvent::KeyboardInput{ input, ..} => match input {
                        glutin::KeyboardInput { virtual_keycode, ..} => match virtual_keycode{
                            Some(x) => println!("{:?}", x),
                            _ => ()
                        },
                    },
                    _ => ()
                },
                _ => ()
            }
        });

        unsafe {
            gl::Clear(gl::COLOR_BUFFER_BIT);
        }

        gl_window.swap_buffers().unwrap();
    }
}
```

**Comment by daboross** [comment]
I'm in a similar situation with ubuntu 17.10 - my files were in `/usr/lib/x86_64-linux-gnu/` rather than `/usr/lib`, but linking `libwayland-egl.so.1` to `libwayland-egl.so` fixed the first error.

For the second error I'm getting `GlutinCreationError(NotSupported)` as well - wrapped because I'm doing it through glium. Looking at what `GlutinCreationError` is, I think I'm running into https://docs.rs/glutin/0.10.1/glutin/enum.CreationError.html#variant.NotSupported.

If I could figure out a way to update wayland-client from 0.9 to 0.12 we could see if this was a bug already fixed, but I'm not sure enough of the internals of glutin to do that.

**Comment by vberger** [comment]
The second issue (the `NotSupported` error) is a purely OpenGL issue, and has nothing to do with wayland-client (and thus I cannot really help you with it, sorry).

The update to wayland-client 0.12 will come once winit 0.9 is out and glutin is updated to use it.

**Comment by daboross** [comment]
Fully reasonable, thank you for your help so far!

If anyone else with more knowledge of glutin's OpenGL side is looking into this, I've tried a few other things. If I run just winit with the example in https://github.com/tomaka/winit README, I get a "phantom" window which doesn't appear on the screen, can't be selected, but does show up in alt+tab.

This is the code I'm using:

```
extern crate winit;

fn main() {
    let mut events_loop = winit::EventsLoop::new();
    let window = winit::Window::new(&events_loop).unwrap();

    window.set_title("Winit Test");
    window.set_inner_size(640, 480);
    window.show();

    events_loop.run_forever(|event| {
        match event {
            winit::Event::WindowEvent { event: winit::WindowEvent::Closed, .. } => {
                winit::ControlFlow::Break
            },
            _ => winit::ControlFlow::Continue,
        }
    });
}
```

Using both winit 0.7 and winit 0.8, I get the following result: (a window I can't actually select, or open, or see, but does "exist" in the window manager)

![phantom-window](https://user-images.githubusercontent.com/1152146/33162505-a68bbabc-cfde-11e7-8343-98c5992bef77.png)

**Comment by vberger** [comment]
This is expected for these versions of winit. Du to how wayland works and how winit uses it, no window appears if you don't draw anything in it.

Note that this will change with next winit version, but I don't know when it'll be released.

**Comment by Tacolizard** [comment]
the solution is actually very simple. i solved the same error in the original issue above by doing `sudo apt-get install libegl1-mesa-dev`

**Comment by daboross** [comment]
Installing `libegl1-mesa-dev` fixes it for me - that seems super strange though. Why would a `-dev` library be required for _running_ a program?

I mean it makes no difference whether or not it's installed while I compile, a binary compiled without `libegl1-mesa-dev` will run correctly if I have the package installed when it's run. I also had `libegl1-mesa` installed before from system dependencies, and that didn't work: only the development library works (and it is `libegl1-mesa-dev` which makes the difference, not any of its dependencies).

**Comment by jedahan** [comment]
Looks like that package not only installs headers but shared libraries

```
/usr/lib/x86_64-linux-gnu/libEGL.so
/usr/lib/x86_64-linux-gnu/libwayland-egl.so
```

(from https://packages.debian.org/sid/amd64/libegl1-mesa-dev/filelist )

maybe that's why?

**Comment by MDeiml** [comment]
For me `/usr/lib/x86_64-linux-libEGL.so` was located at `/usr/lib/x86_64-linux-libEGL.so.1.0.0`. Creating as symlink for that fixed the problem for me, but now my window header looks all messed up.
![screenshot](https://user-images.githubusercontent.com/10165741/37149858-da1df1f8-22cf-11e8-9278-9caa4f819222.png)

**Comment by anderejd** [comment]
> the solution is actually very simple. i solved the same error in the original issue above by doing sudo apt-get install libegl1-mesa-dev

I can confirm that this is required to use glutin on a fresh Ubuntu 18.04 as well.

**Comment by anderejd** [comment]
Should this issue be reopened to investigate this further? It seems weird to need dev packages in order to run the application.

**Comment by francesca64** [comment]
Fixed via #1075
